### DISCUSSION TITLE
Is it possible to have non-standard header file extensions?

### DISCUSSION THREAD

**Discussion by ljden** [clangd/clangd#1655] [Q&A]
I would like to be able to have template implementation header files have the extension `.tpp`. 
ie:
```
// foobar.hpp
#pragma once
template<typename T>
struct foobar {
    void func(T);
}

#include foobar.tpp
```
```
// foobar.tpp
template<typename T>
void foobar<T>::func(T var) {
...
}
```

If I have the file as `foobar.tpp` I get compilation errors. If I change it to `foobar_t.hpp` then clangd is able to give proper diagnostics (so the issue is the file extension).

This is a similar issue to https://github.com/clangd/clangd/issues/1379

I cannot use the pre-compile workaround suggested on [S.O.](https://stackoverflow.com/questions/71471967/clangd-treat-files-with-non-standard-extensions-as-headers) as the file is non-trivial to pre-compile as it relies on system configuration.
I cannot use [another suggestion](https://davy.ai/clangd-treat-files-with-non-standard-extensions-as-headers/) of adding the file with no command as that results with an invalid AST "Could not build CompilerInvocation for file foobar.tpp"

It seems like it may not be currently possible given header extensions are hard coded: https://searchfox.org/llvm/rev/HEAD/clang/lib/Sema/SemaCodeComplete.cpp#10032-10038

Have I missed something? Is this currently possible or would it need to be added to the list of hard coded header extensions? Would it be possible to have a "user-defined" header extension in that list?

**Comment by HighCommander4** [comment]
> If I have the file as `foobar.tpp` I get compilation errors.

Can you show what errors you get?

> It seems like it may not be currently possible given header extensions are hard coded: https://searchfox.org/llvm/rev/HEAD/clang/lib/Sema/SemaCodeComplete.cpp#10032-10038

This list is only relevant for code completion in an `#include` directive (only files with the listed headers will be offered as filename completion proposals).
